core/src/chain_id.rs

```diff
@@ -0,0 +1,185 @@
+    )
+    .unwrap();
+    let time_format =
+        format_description!("[year]-[month]-[day] [hour]:[minute]:[second].[subsecond digits:6]Z");
+    let input = format!(
+        "{}{}{}{}{}{}",
+        constants.k.to_string(),
+        constants.slots_per_epoch.to_string(),
+        constants.slots_per_sub_window.to_string(),
+        constants.delta.to_string(),
+        tx_pool_max_size.to_string(),
+        genesis_timestamp.format(&time_format).unwrap(),
+    );
+    hasher.update(input.as_bytes());
+    hasher.finalize().try_into().unwrap()
+}
+
+impl ChainId {
+    pub fn compute(
+        constraint_system_digests: &[Md5],
+        genesis_state_hash: &StateHash,
+        genesis_constants: &MinaBaseProtocolConstantsCheckedValueStableV1,
+        protocol_transaction_version: u8,
+        protocol_network_version: u8,
+        tx_max_pool_size: &UnsignedExtendedUInt32StableV1,
+    ) -> ChainId {
+        let mut hasher = Blake2b256::default();
+        let constraint_system_hash = constraint_system_digests
+            .iter()
+            .map(|md5| u8_slice_as_hex(md5))
+            .collect::<Vec<String>>()
+            .join("");
+        let genesis_constants_hash = hash_genesis_constants(genesis_constants, tx_max_pool_size);
+        let input = format!(
+            "{}{}{}{}{}",
+            genesis_state_hash,
+            constraint_system_hash,
+            u8_slice_as_hex(&genesis_constants_hash),
+            md5_hash(protocol_transaction_version),
+            md5_hash(protocol_network_version)
+        );
+        hasher.update(input.as_bytes());
+        ChainId(hasher.finalize().try_into().unwrap())
+    }
+
+    pub fn as_hex(&self) -> String {
+        format!("{}", self)
+    }
+
+    pub fn from_hex(s: &str) -> Result<ChainId, ParseIntError> {
+        let mut bytes = [0u8; 32];
+        for i in 0..32 {
+            bytes[i] = u8::from_str_radix(&s[i * 2..i * 2 + 2], 16)?;
+        }
+        Ok(ChainId(bytes))
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> ChainId {
+        let mut arr = [0u8; 32];
+        arr.copy_from_slice(&bytes[..32]);
+        ChainId(arr)
+    }
+}
+
+impl Serialize for ChainId {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        serializer.serialize_str(&self.as_hex())
+    }
+}
+
+impl<'de> Deserialize<'de> for ChainId {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let s = String::deserialize(deserializer)?;
+        ChainId::from_hex(&s).map_err(serde::de::Error::custom)
+    }
+}
+
+impl AsRef<[u8]> for ChainId {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl Display for ChainId {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        for b in self.0 {
+            write!(f, "{:02x}", b)?;
+        }
+        Ok(())
+    }
+}
+
+impl Debug for ChainId {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "ChainId({})", self)
```

Comment:
I would suggest default Debug. Why omit it?

Reply:
I guess hex will be more readable than something the compiler might derive for a `u8` array,